### PR TITLE
Check output of simulations

### LINKED ISSUE #1129: Production tool seems broken
Running the production tool reports a successful run but no simulation is actually performed. I didn't look into it deeply yet, but at minimum I noticed so far that
- no telescopes are actually defined in CORSIKA. 
- There's an error due to zeroth telescope line below:
https://github.com/gammasim/simtools/blob/6484aa889c5fa196ecf6e7ec871c7ec568afeeee/simtools/simtel/simtel_config_writer.py#L179
- The names of the submission scripts contain None.
- etc.

I will try to fix these issues soon, but I thought I should open this issue already to make sure everyone is aware.

P.S.
I discovered these issues because I am writing an integration test that will discover such cases, so at least we should be covered in the future.

### PR DESCRIPTION
**Pull request by orelgueta**
The goal of this PR is to introduce an additional step to the integration tests of the simulation productions that checks the output and makes sure it is reasonable.

The goal above required various changes throughout the code: 
- allow setting random seeds in order to produce reproducible results
- add parameters and schemas that are necessary to reproduce the Prod6 results
- change hardcoded values that did not comply with the Prod6 model

There are also a few random bits of changes introduced as a part of this PR which are not directly related to the subject, but I tried to make sure the commit messages clarify those in case they need clarification.

This PR is dependent on the [corresponding MR](https://gitlab.cta-observatory.org/cta-science/simulations/simulation-model/model_parameters/-/merge_requests/60) in the model parameters repo. Tests will fail until that one is merged.

Fixes #1129.

### CODE REVIEW

File: simtools/corsika/corsika_config.py
Comment (GernotMaier) [inline simtools/corsika/corsika_config.py:56]: This is correct but is without reading the functions below very confusing. Not nice that an additional comment is needed.

File: simtools/model/model_parameter.py
Comment (GernotMaier) [inline simtools/model/model_parameter.py:1]: Do you actually remember why we had this 'or' statement? There must have been a reason.

File: simtools/model/array_model.py
Comment (GernotMaier) [inline simtools/model/array_model.py:114]: I don't like the 'you' in the message - it implies that the user is a person. Maybe "Possibly missing valid layout name or missing telescope list"?

File: simtools/model/site_model.py
Comment (GernotMaier) [inline simtools/model/site_model.py:85]: Good. This closes #829

File: simtools/schemas/model_parameters/save_pe_with_amplitude.schema.yml
Comment (GernotMaier) [inline simtools/schemas/model_parameters/save_pe_with_amplitude.schema.yml:15]: Would `type: boolean` be more appropriate? We have done that with all (most?) other parameters with this type of range (I think Konrad uses this because bool was not available in his C version when he started sim_telarray).

File: simtools/schemas/model_parameters/save_pe_with_amplitude.schema.yml
Comment (GernotMaier) [inline simtools/schemas/model_parameters/save_pe_with_amplitude.schema.yml:16]: It this unit correct? I understood it as boolean.

File: simtools/schemas/model_parameters/store_photoelectrons.schema.yml
Comment (GernotMaier) [inline simtools/schemas/model_parameters/store_photoelectrons.schema.yml:19]: Is this unit correct?

File: simtools/simtel/simulator_array.py
Comment (GernotMaier) [inline simtools/simtel/simulator_array.py:118]: Can you explain the choice of having those values hardwired? Any easy way to make them configurable?

File: simtools/testing/assertions.py
Comment (GernotMaier) [inline simtools/testing/assertions.py:50]: What does `#` mean?

Comment (GernotMaier) [reply simtools/testing/assertions.py:50]: And maybe say that you look into a log file of corsika or simtelarray here?

**Review by GernotMaier — APPROVED** [review]
Thanks a lot for this big update - really hard to keep up with all those little details, I wouldn't have known them.

My comments are mostly minor, this is why I approve already.

**Comment by GernotMaier** [comment]
@orelgueta - I guess you have seen the sonarqu complaint about duplications:

https://sonar-cta-dpps.zeuthen.desy.de/component_measures?id=gammasim_simtools_AY_ssha9WiFxsX-2oy_w&metric=new_duplicated_lines_density&pullRequest=1139&view=list

File: simtools/corsika/corsika_config.py
Comment (orelgueta) [reply simtools/corsika/corsika_config.py:56]: What do you suggest we do then? The other option I see to make it more readable is to change to
```
self.primary_particle = {"primary_id_type": args_dict.get("primary_id_type"), "primary": args_dict.get("primary")}
```

But I think this is worse than the comment explaining it to future developers.

File: simtools/model/array_model.py
Comment (orelgueta) [reply simtools/model/array_model.py:114]: No problem.

File: simtools/schemas/model_parameters/save_pe_with_amplitude.schema.yml
Comment (orelgueta) [reply simtools/schemas/model_parameters/save_pe_with_amplitude.schema.yml:15]: Changed to boolean

File: simtools/schemas/model_parameters/save_pe_with_amplitude.schema.yml
Comment (orelgueta) [reply simtools/schemas/model_parameters/save_pe_with_amplitude.schema.yml:16]: changed to boolean

File: simtools/schemas/model_parameters/store_photoelectrons.schema.yml
Comment (orelgueta) [reply simtools/schemas/model_parameters/store_photoelectrons.schema.yml:19]: No, it's in p.e., but that's not a valid unit, so I set it to dimensionless.

File: simtools/simtel/simulator_array.py
Comment (orelgueta) [reply simtools/simtel/simulator_array.py:118]: I hardcoded them in the same way they are hardcoded in sim_telarray. Since we do not use the sim_telarray histograms, I didn't see a reason to make it configurable. However, if you think it's worth it, please make this comment into an issue and I will implement it after this PR (and use this function to set default values).

File: simtools/testing/assertions.py
Comment (orelgueta) [reply simtools/testing/assertions.py:50]: `#` means number and I think I wrote it because I didn't want to repeat myself with the word "number". I changed it back to number.
This is one of those cases in which I had to reformulate because ruff wouldn't let me have one long sentence which took more than one line...


I do not look into the log files of CORSIKA and sim_telarray here but instead read the sim_telarray output file itself and extract the headers and count the `#` of events and energies produced.

File: simtools/model/model_parameter.py
Comment (orelgueta) [reply simtools/model/model_parameter.py:1]: Yes, it was there in case no label was provided to the application. That caused an issue actually because only in the case of the model parameter the output went to a directory with the name of the telescope instead of a directory under the date (`d-19-09-24`). Then the application would fail because the model parameters were not found. 

This fixes that (though I know you don't like the date thing.

**Comment by orelgueta** [comment]
Thanks for the detailed review! 

I fixed the comments and answered a couple of points. I won't merge yet because of the point regarding the primary particle.

I also fixed the duplication, so I will wait for sonarcube to run again. I didn't know it is so pedantic, but it's not bad I guess.

**Comment by orelgueta** [comment]
The integration tests are failing because of the change to the pixel rate that is missing in the db. The old style of the long string doesn't work anymore, it is replaced with a real list of values instead (which is shortened in the cfg file). We could wait for the model parameter MR to create a new DB, rerun the tests and only then merge this one. That would be the safest.

File: simtools/corsika/corsika_config.py
Comment (GernotMaier) [reply simtools/corsika/corsika_config.py:56]: No suggestion - I was just complaining. Should have made it clearer that no changes are expected.

File: simtools/model/model_parameter.py
Comment (GernotMaier) [reply simtools/model/model_parameter.py:1]: No. I don’t like the date thing - `use_plain_output_path` is my friend. Thanks for the explanation!

File: simtools/simtel/simulator_array.py
Comment (GernotMaier) [reply simtools/simtel/simulator_array.py:118]: I think it is easy to miss that this is for histogramming only. Could we either change the function call or at least say in the first line of the docstring `..used for histogramming in sim_telarray`? Quick readers might otherwise thing we hardwire the spectral index for the simulations.

File: simtools/testing/assertions.py
Comment (GernotMaier) [reply simtools/testing/assertions.py]: Still don’t understand why you can’t use `number`. The first line is not that long.

**Comment by GernotMaier** [comment]
And yes, sonarq is pedantic and our ruff settings are strict. This is good as it really simplifies reviews and we are not bothered with this stuff.

File: simtools/schemas/model_parameters/sampled_output.schema.yml
Comment (GernotMaier) [inline simtools/schemas/model_parameters/sampled_output.schema.yml:12]: Should this be boolean?

File: simtools/schemas/model_parameters/sampled_output.schema.yml
Comment (GernotMaier) [inline simtools/schemas/model_parameters/sampled_output.schema.yml:13]: Should this be null?

File: simtools/simtel/simulator_array.py
Comment (orelgueta) [reply simtools/simtel/simulator_array.py:118]: Changed the name of the function to get_power_law_for_sim_telarray_histograms.

File: simtools/testing/assertions.py
Comment (orelgueta) [reply simtools/testing/assertions.py]: I changed it already to be `Assert the number of showers and the energy range.`

My complaint was about the original version you didn't see which was 
```
Assert the number of showers is consistent with the required one (up to 1% tolerance) 
and that the energies simulated are within the configured ones.
```

This is too long to be one line and ruff doesn't allow a summary line with a line break in it, so I had to break it into two separate statements with some repetition. I find it a bit annoying, so I complained.

File: simtools/schemas/model_parameters/sampled_output.schema.yml
Comment (orelgueta) [reply simtools/schemas/model_parameters/sampled_output.schema.yml:12]: Yes, changed to boolean, I missed it yesterday.

File: simtools/schemas/model_parameters/sampled_output.schema.yml
Comment (orelgueta) [reply simtools/schemas/model_parameters/sampled_output.schema.yml:13]: Removed since it is a boolean

**Comment by orelgueta** [comment]
> Two minor comments, but as before already approved.
> 
> This is what I usually do for testing (and you might have done it):
> 
> * upload the branch in question of the model parameter repository to a local database
> * Run the integration tests locally (remember to run them with `—model_version=“6.0.0”` and `5.0.0`
> 
> If this is fine, we probably can first merge the model_parameter changes; release a new version there (to trigger the DB upload) and then re-run the integration tests here on github.

Yes I have done all of that locally during the development process. Otherwise there's no way to test anything. It should all work now after the merging of both branches.

**Comment by orelgueta** [comment]
I resolved the conflicts with the main branch. I will wait for the tests to pass (or perhaps fail if the DB is not updated yet) and then merge.

**Pull request merged**
